data_structures/priority_queue.h

```diff
@@ -0,0 +1,138 @@
+    queue() {
+        queueFront = NULL;
+        queueRear = NULL;
+        size = 0;
+    }
+
+    /** Destructor */
+    ~queue() = default;
+
+    /** Determine whether the queue is empty */
+    bool isEmptyQueue() { return (queueFront == NULL); }
+
+    /** Add new item to the queue. */
+    void insert(Kind item, Priority lvl) {
+        auto *newNode = new node<Kind, Priority>;
+        newNode->data = item;
+        newNode->next = NULL;
+        newNode->prior = lvl;
+        if (queueFront == NULL) {
+            queueFront = newNode;
+            queueRear = newNode;
+        } else {
+            queueRear->next = newNode;
+            queueRear = queueRear->next;
+        }
+        size++;
+    }
+
+    /** Return the data of the first element of the queue */
+    Kind frontData() {
+        assert(queueFront != NULL);
+        return queueFront->data;
+    }
+
+    /** Return the  first element of the queue */
+    node<Kind, Priority> *front() { return queueFront; }
+
+    /** Return the  last element of the queue */
+    node<Kind, Priority> *last() { return queueRear; }
+
+    /**
+     * Remove the element from the queue that has the highest priority
+     * and return it.
+     */
+    node<Kind, Priority> *deQueue() {
+        if (!isEmptyQueue()) {
+            size--;
+            node<Kind, Priority> *temp = queueFront;
+            auto prev = new node<Kind, Priority>;
+            auto maxPrev = new node<Kind, Priority>;
+            node<Kind, Priority> *after = queueFront->next;
+            Priority max = queueFront->prior;
+            node<Kind, Priority> *maxElem = queueFront;
+            /** Find the highest priority element */
+            while (temp != NULL && (temp == queueRear || temp->next != NULL)) {
+                if (temp->prior > max) {
+                    max = temp->prior;
+                    maxPrev = prev;
+                    maxElem = temp;
+                    after = temp->next;
+                }
+                prev = temp;
+                temp = temp->next;
+            }
+
+            if (maxElem == queueFront) {
+                queueFront = queueFront->next;
+            } else if (maxElem == queueRear) {
+                maxPrev->next = NULL;
+                queueRear = maxPrev;
+            } else if (maxPrev != NULL) {
+                maxPrev->next = after;
+                if (after == NULL)
+                    queueRear = maxPrev;
+            }
+            maxElem->next = NULL;
+            return maxElem;
+        } else {
+            std::cout << "Queue is empty !" << std::endl;
+            return NULL;
+        }
+    }
+
+    /** Clear queue */
+    void clear() { queueFront = NULL; }
+
+    /** Get the size of the queue*/
+    int getSize() { return size; }
+
+ private:
+    node<Kind, Priority> *queueFront; /**< Pointer to the front of the queue */
+    node<Kind, Priority> *queueRear;  /**< Pointer to the rear of the queue  */
+    int size;
```

Comment:
Missing variable documentation.

Reply:
I'm not sure what to add here. This is just the *invariant that keeps track the size of the queue*. Should I add this description as in-line comment?

Reply:
Sure, sounds good. 👍